hypothesis-python/src/hypothesis/extra/pytestplugin.py

```diff
@@ -146,32 +146,46 @@ def pytest_runtest_call(item):
                     raise InvalidArgument(message % (name,))
             yield
         else:
-            # Warn about function-scoped fixtures, excluding autouse fixtures because
-            # the advice is probably not actionable and the status quo seems OK...
-            # See https://github.com/HypothesisWorks/hypothesis/issues/377 for detail.
-            msg = (
-                "%s uses the %r fixture, which is reset between function calls but not "
-                "between test cases generated by `@given(...)`.  You can change it to "
-                "a module- or session-scoped fixture if it is safe to reuse; if not "
-                "we recommend using a context manager inside your test function.  See "
-                "https://docs.pytest.org/en/latest/fixture.html#sharing-test-data "
-                "for details on fixture scope."
-            )
-            argnames = None
-            for fx_defs in item._request._fixturemanager.getfixtureinfo(
-                node=item, func=item.function, cls=None
-            ).name2fixturedefs.values():
-                if argnames is None:
-                    argnames = frozenset(signature(item.function).parameters)
-                for fx in fx_defs:
-                    if fx.argname in argnames:
-                        active_fx = item._request._get_active_fixturedef(fx.argname)
-                        if active_fx.scope == "function":
-                            note_deprecation(
-                                msg % (item.nodeid, fx.argname),
-                                since="2020-02-29",
-                                has_codemod=False,
-                            )
+            # Retrieve the settings for this test from the test object, which
+            # is normally a Hypothesis wrapped_test wrapper. If this doesn't
+            # work, the test object is probably something weird
+            # (e.g a stateful test wrapper), so we skip the function-scoped
+            # fixture check.
+            settings = getattr(item.obj, "_hypothesis_internal_use_settings", None)
+
+            # Check for suspicious use of function-scoped fixtures, but only
+            # if the corresponding health check is not suppressed.
+            if (
+                settings is not None
+                and HealthCheck.function_scoped_fixture
+                not in settings.suppress_health_check
+            ):
```

Comment:
I'm concerned that this will only work for direct application of the `@settings(...)` decorator, whereas I'd want to support this via settings profiles like all other settings.

Can you add a test demonstrating that the profiles approach works?  If it doesn't currently, I think the stash-a-message approach you mentioned in #2733 should do the trick.

Reply:
In theory it should work, since at this point `given` has already attached the default settings to the wrapper function if necessary.

So I'll start by trying to back up my theory with a test that profiles work as expected.

Reply:
It occurs to me that I could also add a hidden `_settings` field to `HypothesisHandle`, and then pull the canonical settings from `item.obj.hypothesis._settings` instead.

The effect is the same, but it's more robust against mistakes in future Hypothesis changes.

(The potential downside is that users could start doing cheeky unsupported things with `_settings` themselves, but I guess there's nothing stopping them from doing similar mischief right now.)

Reply:
(I could instead add a supported `settings` field, but that's too much scope creep for now, especially without a concrete use case.)

Reply:
I'm happy with the current implementation - the important part is that we're happy with the public API; if we want to change the internals after our major release we can do that at any time.

So if you're also happy, let's merge!